packages/core/test/input.test.ts

```diff
@@ -0,0 +1,182 @@
+  });
+
+  it('Should fail when you do pass a property that input does not accept', () => {
+    createMachine<
+      {},
+      AnyEventObject,
+      any,
+      {
+        input: {
+          foo: string;
+        };
+      }
+    >(
+      {},
+      {
+        input: {
+          foo: 'foo',
+          // @ts-expect-error
+          bar: 'bar'
+        }
+      }
+    );
+  });
+
+  it(`Should allow updating the input during the machine's lifecycle with service.input`, (done) => {
+    const machine = createMachine<
+      {},
+      AnyEventObject,
+      any,
+      {
+        input: {
+          foo: string;
+        };
+      }
+    >(
+      {
+        initial: 'notReceivedInput',
+        states: {
+          notReceivedInput: {
+            always: [
+              {
+                cond: (_, __, meta) => {
+                  return meta.state.input.foo === 'bar';
+                },
+                target: 'receivedInput'
+              }
+            ]
+          },
+          receivedInput: {
+            entry: [
+              (_, __, meta) => {
+                expect(meta.state.input).toEqual({
+                  foo: 'bar'
+                });
+                done();
+              }
+            ]
+          }
+        }
+      },
+      {
+        input: {
+          foo: 'foo'
+        }
+      }
+    );
+
+    const service = interpret(machine).start();
+
+    service.input({
+      foo: 'bar'
+    });
+  });
+
+  it('Should allow you to update the input with withConfig', (done) => {
+    const machine = createMachine({
+      entry: [
+        (_, __, meta) => {
+          expect(meta.state.input).toEqual({
+            foo: 'foo'
+          });
+          done();
+        }
+      ]
+    });
+
+    interpret(
+      machine.withConfig({
+        input: {
+          foo: 'foo'
+        }
+      })
+    ).start();
+  });
+
+  describe('When users listen for the input on xstate.input', () => {
+    it('Should already be updated by the time they receive it', (done) => {
+      const machine = createMachine<
+        {},
+        AnyEventObject,
+        any,
+        {
+          input: {
+            foo: string;
+          };
+        }
+      >({
+        on: {
+          'xstate.input': {
+            actions: [
+              (_, __, meta) => {
+                expect(meta.state.input).toEqual({
+                  foo: 'bar'
+                });
+                done();
+              }
+            ]
+          }
+        }
+      });
+
+      const service = interpret(machine).start();
+
+      service.input({
+        foo: 'bar'
+      });
+    });
+  });
+});
+
+describe('createModel', () => {
+  it('Should allow you to define the input type', () => {
+    const model = createModel({}).withInput(
+      {} as {
+        foo: string;
+      }
+    );
+
+    model.createMachine(
+      {},
+      {
+        input: {
+          // @ts-expect-error
+          bar: ''
+        }
+      }
+    );
+  });
+});
```

Comment:
Can we add a test for partial input? If the initial input is:

```js
input: {
  foo: 'bar',
  count: 0
}
```

We should be able to send `service.input({ count: 9 })` and have the resulting input be:

```js
input: {
  foo: 'bar',
  count: 9
}
```